src/firmware/app/world/firmware_robot.c

```diff
@@ -119,32 +261,313 @@ float app_firmware_robot_getBatteryVoltage(const FirmwareRobot_t* robot)
 }
 
-RobotConstants_t app_firmware_robot_getRobotConstants(const FirmwareRobot_t* robot)
+float app_firmware_robot_getBackLeftWheelSpeedRPM(const FirmwareRobot_t* robot)
 {
-    return robot->robot_constants;
+    return app_force_wheel_getMotorSpeedRPM(robot->back_left_force_wheel);
 }
 
-ControllerState_t* app_firmware_robot_getControllerState(const FirmwareRobot_t* robot)
+float app_firmware_robot_getBackRightWheelSpeedRPM(const FirmwareRobot_t* robot)
 {
-    return robot->controller_state;
+    return app_force_wheel_getMotorSpeedRPM(robot->back_right_force_wheel);
+}
+
+void app_firmware_robot_trackVelocityInRobotFrame(const FirmwareRobot_t* robot,
+                                                  float linear_velocity_x,
+                                                  float linear_velocity_y,
+                                                  float angular_velocity)
+{
+    ForceWheel_t* front_right_wheel = robot->front_right_force_wheel;
+    ForceWheel_t* front_left_wheel  = robot->front_left_force_wheel;
+    ForceWheel_t* back_right_wheel  = robot->back_right_force_wheel;
+    ForceWheel_t* back_left_wheel   = robot->back_left_force_wheel;
+
+    const RobotConstants_t robot_constants = app_firmware_robot_getRobotConstants(robot);
+    ControllerState_t* controller_state    = app_firmware_robot_getControllerState(robot);
+    float current_vx                       = app_firmware_robot_getVelocityX(robot);
+    float current_vy                       = app_firmware_robot_getVelocityY(robot);
+    float current_angular_velocity         = app_firmware_robot_getVelocityAngular(robot);
+    float current_orientation              = app_firmware_robot_getOrientation(robot);
+    float battery_voltage                  = app_firmware_robot_getBatteryVoltage(robot);
+
+    // Rotate the current_velocity vector from the world frame to the robot frame
+    float current_velocity[2];
+    current_velocity[0] = current_vx;
+    current_velocity[1] = current_vy;
+    rotate(current_velocity, -current_orientation);
+
+    // This is the "P" term in a PID controller. We essentially do proportional
+    // control of our acceleration based on velocity error
+    static const float VELOCITY_ERROR_GAIN = 10.0f;
+
+    float desired_acceleration[2];
+    desired_acceleration[0] =
+        (linear_velocity_x - current_velocity[0]) * VELOCITY_ERROR_GAIN;
+    desired_acceleration[1] =
+        (linear_velocity_y - current_velocity[1]) * VELOCITY_ERROR_GAIN;
+
+    float angular_acceleration =
+        (angular_velocity - current_angular_velocity) * VELOCITY_ERROR_GAIN;
+
+    app_control_applyAccel(robot_constants, controller_state, battery_voltage,
+                           front_left_wheel, front_right_wheel, back_left_wheel,
+                           back_right_wheel, desired_acceleration[0],
+                           desired_acceleration[1], angular_acceleration);
+}
+
+void force_wheels_followPosTrajectory(const FirmwareRobot_t* robot,
```

Comment:
nit: all these functions need the `app_firmware_robot` prefix infront of them

Reply:
Would this be true since we're only calling them from within this file since the differentiation b/w `force_wheels` and `velocity_wheels` is handled when `app_firmware_robot_followPosTrajectory` is called?

Reply:
ooh good point, but these functions are defined in `firmware_robot.c` so other files can still call them

are these supposed to be private helpers? if so you can move the definitions to be in the .c file at the top so only `firmware_robot.c` can use them

Reply:
Done.